crates/sncast/src/lib.rs

```diff
@@ -412,32 +413,41 @@ pub fn print_command_result<T: Serialize>(
     value_format: ValueFormat,
     json: bool,
 ) -> Result<()> {
-    let mut output = vec![("command", command.to_string())];
-    let json_value: Value;
+    let mut output = vec![("command".to_string(), command.to_string())];
 
     let mut error = false;
     match result {
         Ok(result) => {
-            json_value = serde_json::to_value(result)
-                .map_err(|_| anyhow!("Failed to convert command result to serde_json::Value"))?;
-
-            output.extend(
-                json_value
-                    .as_object()
-                    .expect("Invalid JSON value")
-                    .iter()
-                    .filter_map(|(k, v)| value_format.format_json_value(v).map(|v| (k.as_str(), v)))
-                    .collect::<Vec<(&str, String)>>(),
-            );
+            output.extend(stringify_command_result_struct(result, value_format)?);
         }
         Err(message) => {
-            output.push(("error", format!("{message:#}")));
+            output.push(("error".to_string(), format!("{message:#}")));
             error = true;
         }
     };
     print_formatted(output, json, error)
 }
 
+pub fn stringify_command_result_struct<T: Serialize>(
+    command_result: &T,
+    value_format: ValueFormat,
+) -> Result<Vec<(String, String)>> {
+    let json_value = serde_json::to_value(command_result)
+        .map_err(|_| anyhow!("Failed to convert command result to serde_json::Value"))?;
+
+    let result = if let Some(json_object) = json_value.as_object() {
+        json_object
+            .iter()
+            .filter_map(|(k, v)| value_format.format_json_value(v).map(|v| (k.to_owned(), v)))
```

Comment:
what if obtained json is invalid?

Reply:
I changed it in my PR to solve some of those issues, I suggest rebasing this one on that.

Reply:
merged master branch